### PR TITLE
Overridable command changes

### PR DESCRIPTION
**Pull request by NikhilSuresh24**
made changes to overridablecommand stuff, and made changes to master to account for it

### CODE REVIEW

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 1]
```diff
@@ -1,30 +1,23 @@
...
+import edu.wpi.first.wpilibj.command.Command;
 
-	/**
-	 * returns if override is on or off
-	 * 
-	 * @return
-	 */
-	boolean isOverridden();
+public class OverridableCommand extends Command implements Overridable {
+	public boolean override;
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: You should make this private, and call it `isOverridden`.

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 2]
```diff
@@ -1,30 +1,23 @@
...
 
-	/**
-	 * opposite of getOverride
-	 * 
-	 * @return
-	 */
-	default boolean isNotOverridden() {
-		return !isOverridden();
+	@Override
+	public void setOverridden(boolean override) {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: A quick javadoc?

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 3]
```diff
@@ -1,30 +1,23 @@
...
+		this.override = override;
 	}
-}
\ No newline at end of file
+
+	@Override
+	public boolean isOverridden() {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: See above.

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 4]
```diff
@@ -1,30 +1,23 @@
...
-	void setOverride(boolean setValue);
+import edu.wpi.first.wpilibj.command.Command;
 
-	/**
-	 * returns if override is on or off
-	 * 
-	 * @return
-	 */
-	boolean isOverridden();
+public class OverridableCommand extends Command implements Overridable {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: This should be `abstract`, because it should only be used by extension :)

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 5]
```diff
@@ -1,30 +1,23 @@
...
-}
\ No newline at end of file
+
+	@Override
+	public boolean isOverridden() {
+		return override;
+	}
+
+	@Override
+	protected boolean isFinished() {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: Once the class is `abstract` you can remove this method. Abstract classes aren't required to implement all the abstract method from the parent class.

File: src/org/usfirst/frc4904/robot/commands/OverrideableCommandGroup.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+
+import org.usfirst.frc4904.standard.commands.RunIf;
+import edu.wpi.first.wpilibj.command.Command;
+import edu.wpi.first.wpilibj.command.CommandGroup;
+
+public abstract class OverrideableCommandGroup extends CommandGroup implements Overridable {
+	protected volatile boolean isOverriden;
+
+	public OverrideableCommandGroup() {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/OverrideableCommandGroup.java]: These constructors will be auto-inherited—you don't need to re-define them unless you want to add additional code in each.

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java (hunk 2 above)
Comment (ajnadel) [reply src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: Also make these `final public synchronized` for parity with the other class.

File: src/org/usfirst/frc4904/robot/commands/SetOverride.java [hunk 7]
```diff
@@ -5,17 +5,17 @@
 
 public class SetOverride extends Command {
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/SetOverride.java]: Make this extend `InstantCommand` instead `Command` because everything happens in `initialize`.

File: src/org/usfirst/frc4904/robot/commands/SetRampState.java [hunk 8]
```diff
@@ -3,22 +3,18 @@
...
 	protected final RampState state;
+	public static boolean rampOverride;
 
 	public SetRampState(RampState state) {
 		this.state = state;
 	}
 
 	@Override
 	protected void initialize() {
+		override = SetRampState.rampOverride;
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/SetRampState.java]: Use `setOverride()`

File: src/org/usfirst/frc4904/robot/humaninterface/operators/DefaultOperator.java [hunk 9]
```diff
@@ -33,9 +33,9 @@ public void bindCommands() {
 		new ThresholdCommand(new GearioOuttake(), RobotMap.Component.operatorStick::getY, -DefaultOperator.INTAKE_THRESHOLD,
 			true)
 				.start();
-		RobotMap.Component.gearIO.setRampState(GearIO.RampState.EXTENDED);
-		RobotMap.Component.teensyStick.getButton(11).whenPressed(new SetOverride(true, RobotMap.Component.gearIO));
-		RobotMap.Component.teensyStick.getButton(11).whenReleased(new SetOverride(false, RobotMap.Component.gearIO));
+		new SetRampState(GearIO.RampState.EXTENDED).start();
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/humaninterface/operators/DefaultOperator.java]: Style nit: move all uncontained commands to the end of `bindCommands()`

File: src/org/usfirst/frc4904/robot/commands/SetRampState.java [hunk 10]
```diff
@@ -3,22 +3,18 @@
...
 		RobotMap.Component.gearIO.setRampState(state);
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/commands/SetRampState.java]: Also, this will happen regardless of the override—you have to explicitly check it.

File: src/org/usfirst/frc4904/robot/subsystems/GearIO.java [hunk 11]
```diff
@@ -1,27 +1,28 @@
...
-	public boolean rampOverride;
 
 	public GearIO(Motor intakeRoller, DoubleSolenoid gullWings, DoubleSolenoid ramp) {
 		this.intakeRoller = intakeRoller;
 		this.gullWings = gullWings;
 		this.ramp = ramp;
 		setState(GearState.TRANSPORT);
-		setRampState(GearIO.RampState.EXTENDED);
+		if (new SetRampState(null).isNotOverridden()) {
+			setRampState(GearIO.RampState.EXTENDED);
```
Comment (ajnadel) [inline src/org/usfirst/frc4904/robot/subsystems/GearIO.java]: This is incredibly janky and I won't have it.

Comment (ajnadel) [reply src/org/usfirst/frc4904/robot/subsystems/GearIO.java]: Let's just make the override on a command-by-command basis and not on a static level.

File: src/org/usfirst/frc4904/robot/commands/SetRampState.java (hunk 8 above)
Comment (NikhilSuresh24) [reply src/org/usfirst/frc4904/robot/commands/SetRampState.java]: we need the override to apply to all setRampStates but we cant make the override in overridable command static, or else it will apply to all overridable commands

File: src/org/usfirst/frc4904/robot/commands/SetOverride.java (hunk 7 above)
Comment (NikhilSuresh24) [inline src/org/usfirst/frc4904/robot/commands/SetOverride.java]: it should be a overridable command

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 12]
```diff
@@ -1,30 +1,29 @@
...
-	/**
-	 * sets a static override to a boolean you give it
-	 * 
-	 * @param override
-	 */
-	void setOverride(boolean setValue);
+import edu.wpi.first.wpilibj.command.Command;
+
+public abstract class OverridableCommand extends Command implements Overridable {
+	private volatile boolean isOverridden;
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: Why is this volatile? There should not be any threading involved in overriables.

File: src/org/usfirst/frc4904/robot/commands/OverridableCommand.java [hunk 13]
```diff
@@ -1,30 +1,29 @@
...
-	 * opposite of getOverride
+	 * Get whether this command is overridden.
 	 * 
-	 * @return
+	 * @returns Whether this command is overridden.
 	 */
-	default boolean isNotOverridden() {
-		return !isOverridden();
+	@Override
+	final public synchronized boolean isOverridden() {
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/OverridableCommand.java]: See above.

File: src/org/usfirst/frc4904/robot/commands/SetOverride.java (hunk 7 above)
Comment (carturn) [reply src/org/usfirst/frc4904/robot/commands/SetOverride.java]: I agree.

File: .classpath [hunk 14]
```diff
@@ -13,7 +13,6 @@
 	<classpathentry kind="var" path="opencv" sourcepath="opencv.sources"/>
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib-javadoc.jar"/>
-	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib-sources.jar"/>
```
Comment (carturn) [inline .classpath]: This line should not be removed.

File: src/org/usfirst/frc4904/robot/commands/Climb.java [hunk 15]
```diff
@@ -23,7 +23,7 @@ protected void execute() {
 		set(Math.max(0, axis.get())); // No, really, always feed positive values
 										// to the climber motor
 		if (axis.get() >= Climb.MINIMUM_AXIS_VALUE_TO_FLIP_GEARIO_RAMP && RobotMap.Component.gearIO.isNotOverridden()) {
-			RobotMap.Component.gearIO.setRampState(GearIO.RampState.RETRACTED);
+			new SetRampState(GearIO.RampState.RETRACTED).start();
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/Climb.java:26]: In a perfect world, this would probably be on a separate pull request, but I do not think that it is too much of a problem.

File: src/org/usfirst/frc4904/robot/commands/Overridable.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package org.usfirst.frc4904.robot.commands;
+
+
+/**
+ * This interface allows commands to easily override other commands
```
Comment (carturn) [inline src/org/usfirst/frc4904/robot/commands/Overridable.java]: Please update javadoc.

**Review by carturn — APPROVED** [review]
I am Carter, and I approve this pull request.

**Comment by c0d3rman** [comment]
Made obsolete by #44
